rpm-ostree-bisect

```diff
@@ -0,0 +1,347 @@
+#
+# Later check systemctl status rpm-ostree-bisect.service for result
+# 
+import argparse
+import json
+import os
+import os.path
+import subprocess
+import sys
+from collections import OrderedDict
+
+import gi
+gi.require_version('OSTree', '1.0')
+from gi.repository import GLib, Gio, OSTree
+
+DATA_FILE = '/var/lib/rpm-ostree-bisect.json'
+SYSTEMD_UNIT_FILE = '/etc/systemd/system/rpm-ostree-bisect.service'
+SYSTEMD_UNIT_NAME = 'rpm-ostree-bisect.service'
+SYSTEMD_UNIT_CONTENTS = """
+[Unit]
+Description=RPM-OSTree Bisect Testing
+After=network.target
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+#ExecStart=/usr/bin/sleep 20
+ExecStart=/usr/local/bin/rpm-ostree-bisect --resume
+StandardOutput=journal+console
+StandardError=journal+console
+
+[Install]
+WantedBy=multi-user.target
+"""
+
+
+def fatal(msg):
+    print(msg, file=sys.stderr)
+    sys.exit(1)
+
+def log(msg):
+    print(msg)
+    sys.stdout.flush()
+
+""" 
+    Initialize commit info ordered dict. The array will be a list of 
+    commits in descending order. Each entry will be a dict with 
+    key of commitid and value = a dict of version, heuristic
+    (TESTED, GIVEN, ASSUMED), and status (GOOD/BAD/UNKNOWN)
+
+    commits = {
+        'abcdef' => { 'version': '28.20180302.0' ,
+                      'heuristic', 'GIVEN',
+                      'status': 'BAD',
+                    },
+        'bbcdef' => { 'version': '28.20180301.0' ,
+                      'heuristic', 'ASSUMED',
+                      'status': 'UNKNOWN',
+                    },
+        'cbcdef' => { 'version': '28.20180228.0' ,
+                      'heuristic', 'TESTED',
+                      'status': 'GOOD',
+                    },
+    }
+"""
+def initialize_commits_info(repo, bad, good):
+    # An ordered dictionary of commit info
+    info = OrderedDict()
+    # The first commit in our list will be the "BAD" commit
+    commitid = bad
+
+    # Iterate over all commits and add them to the array
+    while commitid is not None:
+        _, commit = repo.load_variant(OSTree.ObjectType.COMMIT, commitid)
+        meta = commit.get_child_value(0)
+        version = meta.lookup_value('version', GLib.VariantType.new('s'))
+        info.update({ commitid: { 'version': version.get_string(), 
+                                 'heuristic': 'ASSUMED',
+                                 'status': 'UNKNOWN' }})
+        # Next iteration
+        commitid = OSTree.commit_get_parent(commit)
+
+    # Mark the bad commit bad
+    info[bad]['status'] = 'BAD'
+    info[bad]['heuristic'] = 'GIVEN'
+
+    # Mark the good commit good
+    if good:
+        info[good]['status'] = 'GOOD'
+        info[good]['heuristic'] = 'GIVEN'
+    else:
+        lastcommit = list(info.keys())[-1]
+        info[lastcommit]['status'] = 'GOOD'
+
+    return info
+
+"""
+    Grab all commit history from the remote (just
+    the metadata).
+"""
+def pull_commit_history(deployment, repo):
```

Comment:
No objections to using the API, but it might be simpler to just shell out to `ostree pull`.

Reply:
It's definitely useful to learn it...I guess to me the "API vs shell callout" comes down to "are you doing something hard/different"?  Pulp today uses the OSTree API which makes a ton of sense.   But in this case we're just writing output to the TTY.